### DISCUSSION TITLE
Wait for translation to load and than show translations

### DISCUSSION THREAD

**Discussion by alejandrocoding** [ngneat/transloco#619] [Q&A] [answered]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Which Transloco package(s) are the source of the bug?

Transloco

### Is this a regression?

No

### Current behavior

I have a dropdown language selector, when I switch to another language I show a snackbar.

The problem is that if the language has not been loaded before, the snackbar does not show the translation, but rather the json key path to that value. Like if the key was missing.

Workaround: Using a setTimeout of 100ms before showing the snackbar works.

More thoughts about it:
- To change the language I am using `setActiveLang('lang-code')` and that method is asynchronus, I wonder if it's a better idea to return the promise/observable rather than void.
- Because the workaround is not of my joy, I tried to use `transloco.langChanges$`, subscribes to it and from there call the snackbar with the message. Also fail. Would be a good idea to emit once the file has been fully loaded or expose another observable for that?

This does not affect pipes or directive content, probably given the way they are made.

<img width="363" alt="err" src="https://user-images.githubusercontent.com/5755545/210917511-9ca81d60-8010-409b-af24-05b29c0a377a.png">

### Expected behavior

I would expect there should be a way to be notified once a language is fully loaded and ready to be used correctly.

Note: If I am missing something that I haven't seen in the doc I very apologize in advance.

### Please provide a link to a minimal reproduction of the bug

i-can-try-to-provide-something-later

### Please provide the environment you discovered this bug in

```markdown
Transloco: 
Angular: 
Node: 
Package Manager: 
OS:
```

### I would like to make a pull request for this bug

Yes 🚀

**Comment by shaharkazaz** [comment, marked as answer]
@alejandrocoding I've converted this into a discussion as this is a question relating to a specific use case, if we find that there is a generalized problem here we can address this as an issue :)
Have you tried the [events$](https://ngneat.github.io/transloco/docs/translation-api/#events) observable to indicate when to call your snackbar?

**Reply by alejandrocoding** [reply]
Thanks @shaharkazaz! what a quick response!

Nope, I haven't tried that one, I have tried the langChanged$ which I guess it's the same than:
`this.translocoService.events$.pipe(filter(e => e.type === 'langChanged'))`

Out of curiosity, I am going to try that out. If it doesn't work, I'll add a `delay(100)`.
I'm not happy with that workaround, because if the app is used in a mobile with a poor network connection, it won't work :( - but that's the only workaround I've come with. If you think of any idea to try, please let me know!

Thank you!

PS: I've tried events$ and it's not working.

**Reply by shaharkazaz** [reply]
@alejandrocoding Can you please share a small reproduction so I can better assist? 
I understand your thought regarding these streams but they are not the same, this is actually one of the things I'm taking into consideration when designing the next major version.
I don't want the resolution to be just some arbitrary delay.

**Reply by alejandrocoding** [reply]
There you go @shaharkazaz 
https://stackblitz.com/edit/transloco-lang-bug?file=src/app/app.component.ts

I've included the discussed work arounds (they all are based on an arbitrary delay, but different strategies).

I have checked with some mates using the library to see if they are experiencing the same, and they are not because they need to reload after language changes. I think this is the key part. I see this issue happening to everyone who has a combination of: reRenderOnLangChange: false + show info of action by any component (snackbar, toast, popup...) - and this sounds to me like a wide range of scenarios, but I am not sure how many folks are refreshing after lang changes.

**Reply by shaharkazaz** [reply]
@alejandrocoding You just need to use `selectTranslate` which waits for the translation to load, based on the example provided the following will work:
```ts
    this.translocoService.setActiveLang(lang);
    this.translocoService.selectTranslate('message').pipe(take(1)).subscribe((msg) => {
      this.snackBar.open(msg, '', { duration: 5000 });
    });    
```

**Reply by alejandrocoding** [reply]
I don't know how I missed that part of the documentation where you are making this big and clear!

Probably I was just focused on finding an observable that emitted when ready rather than a method and skipped that part of the doc, I am sorry for taking your time looking at this.

![doc](https://user-images.githubusercontent.com/5755545/211156686-252fc5e8-3ff7-4267-9372-ebfa5538791d.png)

Thanks @shaharkazaz

**Reply by shaharkazaz** [reply]
@alejandrocoding You're welcome.
On a different note, If you or your company find this lib and other ngneat OS useful please consider sponsoring us to keep these projects going 🙏 
We have different perks for sponsorship tiers, you can check them out [here](https://github.com/sponsors/ngneat).
